crates/gen-wasmtime-py/src/source.rs

```diff
@@ -0,0 +1,495 @@
+use heck::*;
+use wit_bindgen_gen_core::wit_parser::*;
+
+use crate::dependencies::Dependencies;
+
+/// A [Source] represents some unit of Python code
+/// and keeps track of its indent.
+#[derive(Default)]
+pub struct Source {
+    s: String,
+    indent: usize,
+}
+
+impl Source {
+    /// Appends a string slice to this [Source].
+    ///
+    /// Strings without newlines, they are simply appended.
+    /// Strings with newlines are appended and also new lines
+    /// are indented based on the current indent level.
+    pub fn push_str(&mut self, src: &str) {
+        let lines = src.lines().collect::<Vec<_>>();
+        let mut trim = None;
+        for (i, line) in lines.iter().enumerate() {
+            self.s.push_str(if lines.len() == 1 {
+                line
+            } else {
+                let trim = match trim {
+                    Some(n) => n,
+                    None => {
+                        let val = line.len() - line.trim_start().len();
+                        if !line.is_empty() {
+                            trim = Some(val);
+                        }
+                        val
+                    }
+                };
+                line.get(trim..).unwrap_or("")
+            });
+            if i != lines.len() - 1 || src.ends_with("\n") {
+                self.newline();
+            }
+        }
+    }
+
+    /// Prints the documentation as comments
+    /// e.g.
+    /// > \# Line one of docs node
+    /// >
+    /// > \# Line two of docs node
+    pub fn comment(&mut self, docs: &Docs) {
+        let docs = match &docs.contents {
+            Some(docs) => docs,
+            None => return,
+        };
+        for line in docs.lines() {
+            self.push_str(&format!("# {}\n", line));
+        }
+    }
+
+    /// Prints the documentation as comments
+    /// e.g.
+    /// > """
+    /// >
+    /// > Line one of docs node
+    /// >
+    /// > Line two of docs node
+    /// >
+    /// > """
+    pub fn docstring(&mut self, docs: &Docs) {
+        let docs = match &docs.contents {
```

Comment:
I'm still learning the rusty patterns and here I wonder what your thoughts are with using 
```rust 
docs.contents.as_ref().map(|docs| {...})
```
Or

```
if let Some(docs) = &docs.contents {
...
}
```
Again just a style choice, but personally like avoiding `returns` in rust and have come to like the latter choice.

Reply:
This is carried over directly from the docstring PR, which itself pulled this code from the original `comment` implementation. I don't have any strong opinions on this myself.

Reply:
Aw makes sense.  Honestly all of them are readable so doesn't really matter.